ompi/mca/pml/cm/pml_cm_component.c

```diff
@@ -106,16 +106,653 @@ mca_pml_cm_component_register(void)
+    ASSERT(0 == opal_accelerator.record_event(&event1, stream));  // Record Event 1
+    ASSERT(0 == opal_accelerator.synchronize_stream(stream));     // Sync Stream, Event 1 should be complete
+    ASSERT(0 == opal_accelerator.record_event(&event2, stream));  // record event 2,
+    ASSERT(0 == opal_accelerator.synchronize_event(&event1));     // sync event 1, already complete
+    ASSERT(0 == opal_accelerator.synchronize_event(&event2));     // sync event 2, should be complete
+    ASSERT(0 == opal_accelerator.synchronize_event(&event3));     // Event 3 was never added to stream, always complete
+
+    // Re-Add Events to stream
+    ASSERT(0 == opal_accelerator.record_event(&event1, stream));
+    ASSERT(0 == opal_accelerator.record_event(&event2, stream));
+    ASSERT(0 == opal_accelerator.record_event(&event3, stream));
+
+    // Sync Stream, should be fine
+    ASSERT(0 == opal_accelerator.synchronize_stream(stream));
+
+    // Re-Add Events to stream
+    ASSERT(0 == opal_accelerator.record_event(&event1, stream));
+    ASSERT(0 == opal_accelerator.record_event(&event2, stream));
+    ASSERT(0 == opal_accelerator.record_event(&event3, stream));
+
+    // Sync Event 3, should block forever
+    ASSERT(0 == opal_accelerator.synchronize_event(&event3));     // Event 3 was never added to stream, always complete
+
+    // Successfully Destroy a stream and events
+    ASSERT(0 == opal_accelerator.destroy_stream(stream));
+    ASSERT(0 == opal_accelerator.destroy_event(&event1));
+    ASSERT(0 == opal_accelerator.destroy_event(&event2));
+    ASSERT(0 == opal_accelerator.destroy_event(&event3));
+
+    printf("TEST PASSED, accelerator_test_stream() succeeded!!!\n");
+}
+
+static void accelerator_test_memcpy_async(int dev_id) {
+    int rv = -1;
+    int size =  sizeof(int) * 1024 * 1024 * 500;
+    int *memory_on_device = NULL, *sending = NULL, *receiving = NULL;
+    opal_accelerator_stream_t *stream = NULL, *stream1 = NULL;
+    opal_accelerator_event_t event = {};
+
+    // Allocate Host Memory
+    sending = (int*) malloc(size);
+    receiving = (int*) malloc(size);
+
+    // Set Host Buffers to initial values
+    for (int i = 0; i < size / sizeof(int); i++) {
+        sending[i] = i;
+        receiving[i] = -1;
+    }
+
+    // Allocate Memory On GPU
+    ASSERT(0 == opal_accelerator.malloc(&memory_on_device, size));
+
+    // Create ASYNC Objects
+    ASSERT(0 == opal_accelerator.create_stream(dev_id, &stream));
+    ASSERT(0 == opal_accelerator.create_stream(dev_id, &stream1));
+    ASSERT(0 == opal_accelerator.create_event(&event));
+
+    // Perform Memcpy from host to device
+    ASSERT(0 == opal_accelerator.memcpy_async(dev_id, memory_on_device, sending, size, stream, MCA_ACCELERATOR_TRANSFER_UNSPEC));
+    // Add Event to Stream
+    ASSERT(0 == opal_accelerator.record_event(&event, stream));
+    // sync event 1
+    ASSERT(0 == opal_accelerator.synchronize_event(&event));
```

Comment:
I commented out the synchronize's and couldn't ever get this to fail... I ran it in a loop 100x and used big transfers so idk why I didn't hit the race condition that I put in there.